### PR TITLE
Ci/precommit hook and unused fix

### PR DESCRIPTION
**Pull request by sago2k8**
# Cover letter 

This issue aims to improve developer experience by automating some check that would happen anyhow in the CI process, so we can find specific issues before pushing to the remote repository, also followed @rikimaru0345  recommendation and extend the unused variable rule to support `_` as valid unused variable.

## Points of interest 

- allow `^_` as a valid unused variable
- extend from `plugin:@typescript-eslint/eslint-recommended` to add some extra linting rules
- delete unused Eslint config
- add a pre-commit hook to pass linter before committing changes to git

### CODE REVIEW

File: frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts [hunk 1]
```diff
@@ -29,45 +29,45 @@ import { toJson } from '../../../../utils/jsonUtils';
...
 export type PartitionAssignments = {
-    partition: Partition,
-    brokers: Broker[] // brokers the replicas are on; meaning length is always equal to the replicationFactor
+    partition: Partition;
+    brokers: Broker[]; // brokers the replicas are on; meaning length is always equal to the replicationFactor
 };
 
-type BrokerReplicaCount = { // track how many replicas were assigned to a broker (over all partitions)
+type BrokerReplicaCount = {
+    // track how many replicas were assigned to a broker (over all partitions)
```
Comment (rikimaru0345) [inline frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts:50]: that comment is intended to describe the whole type, not the first property

File: frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts [hunk 2]
```diff
@@ -29,45 +29,45 @@ import { toJson } from '../../../../utils/jsonUtils';
...
     // Track information like used disk space per broker, so we extend each broker with some metadata
-    const allExBrokers = apiData.brokers.map(b => new ExBroker(b));
-    const targetExBrokers = allExBrokers.filter(exb => targetBrokers.find(b => exb.brokerId == b.brokerId) != undefined);
+    const allExBrokers = apiData.brokers.map((b) => new ExBroker(b));
+    const targetExBrokers = allExBrokers.filter(
+        (exb) => targetBrokers.find((b) => exb.brokerId == b.brokerId) != undefined
+    );
     const getExBroker = (brokerId: number): ExBroker => {
-        const exBroker = allExBrokers.find(x => x.brokerId == brokerId);
+        const exBroker = allExBrokers.find((x) => x.brokerId == brokerId);
```
Comment (rikimaru0345) [inline frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts:70]: which rule is adding parameter parenthesis here?

File: frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts [hunk 3]
```diff
@@ -76,8 +76,7 @@ function computeReassignments(
     // For the sake of calculation, it is easier to start with a fresh slate.
     // So we first 'virtually' remove these assignments by going through each replica
     // and subtracting its size(disk space) from broker it is on.
-    for (const broker of allExBrokers)
-        broker.recompute(apiData, selectedTopicPartitions);
+    for (const broker of allExBrokers) broker.recompute(apiData, selectedTopicPartitions);
```
Comment (rikimaru0345) [inline frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts:79]: this is harder to read than the 2-line version

File: frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts [hunk 4]
```diff
@@ -413,26 +439,31 @@ class ExBroker implements Broker {
...
 
+    get plannedReplicas(): number {
+        return this.initialReplicas + this.assignedReplicas;
+    }
+    get plannedSize(): number {
+        return this.initialSize + this.assignedSize;
+    }
+    get plannedLeader(): number {
+        return this.initialLeader + this.assignedLeader;
+    }
```
Comment (rikimaru0345) [inline frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts:466]: single line version was fine in my opinion

File: frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts [hunk 5]
```diff
@@ -449,23 +480,27 @@ class ExBroker implements Broker {
         this.actualLeader = 0;
 
         if (apiData.topicPartitions == null)
-            throw new Error(`cannot recompute actual usage of broker '${this.brokerId}' because 'api.topicPartitions == null' (no permissions?)`);
+            throw new Error(
+                `cannot recompute actual usage of broker '${this.brokerId}' because 'api.topicPartitions == null' (no permissions?)`
+            );
```
Comment (rikimaru0345) [inline frontend/src/components/pages/reassign-partitions/logic/reassignLogic.ts:485]: added line breaks don't make this easier to read imho

File: frontend/src/utils/interpreter/findFunction.ts [hunk 6]
```diff
@@ -9,37 +9,37 @@
...
 // declare function findAll(isMatch: (obj:object|Array<any>)=>boolean): any[];
 // declare function findAll(pattern: object, ignoreCase?:boolean): object[];
 function findAll(this: any | undefined, arg1: any, arg2: any): any {
-    const self = (this != null)
-        ? this // called on object
-        : value; // called in root
+    const self =
+        this != null
+            ? this // called on object
+            : value; // called in root
```
Comment (rikimaru0345) [inline frontend/src/utils/interpreter/findFunction.ts:42]: Maybe the rule can be configured to leave the condition on the same line ?

File: frontend/src/utils/interpreter/findFunction.ts [hunk 7]
```diff
@@ -50,35 +50,28 @@ function findAll(this: any | undefined, arg1: any, arg2: any): any {
...
     const ignoreCase = Boolean(arg2);
     const caseSensitive = !ignoreCase;
 
     if (typeof arg1 == 'string') {
         // findByName
         const propertyName = String(arg1);
         return findByName(self, propertyName, caseSensitive, returnFirstResult);
-    }
-    else if (typeof arg1 == 'function') {
+    } else if (typeof arg1 == 'function') {
```
Comment (rikimaru0345) [inline frontend/src/utils/interpreter/findFunction.ts:62]: I liked the extra line break. It helped to indicate that a new block is starting.
If we have this rule, I would like it to add a line break before any `else if` blocks.

**Review by rikimaru0345 — CHANGES_REQUESTED** [review]
There are some good changes, but many changes/rules go a bit overboard and (at least in my opinion) make the code less readable.

However most rules probably have options that we can set up to get a net improvement.

File: frontend/src/utils/interpreter/findFunction.ts (hunk 7 above)
Comment (sago2k8) [reply frontend/src/utils/interpreter/findFunction.ts]: I see this, Gonna try to find a way to enforce it. and add it if possible, I think in this case my prettier settings did the change.

**Comment by sago2k8** [comment]
Thanks @rikimaru0345, I update the code and added some more clear and useful rules.
addressed and fixed your comments. could you review it again

**Pull request merged**